### PR TITLE
FEATURE - User Config Validation

### PR DESCRIPTION
**Pull request by dannypaz**
## Description
If a user's custom configuration file is invalid, we want the broker to let the user know that we are using a default configuration, so that they are not confused when their commands are showing different information.

We also provide a tool for users to validate their config, making sure that it is valid javascript.

Users can now use `npm run validate-config` to validate their `~/.sparkswap.js` file.

An examples of warning message:
screen shot 2018-08-07 at 10 31 44 am

An Example of an invalid configuration:
screen shot 2018-08-07 at 10 28 49 am

## Related PRs
List related PRs if applicable

## Todos
- [n/a] Tests
- [x] Documentation
- [x] Link to Trello

### CODE REVIEW

**Comment by dannypaz** [comment]
[Configuration validation for broker-daemon](https://trello.com/c/M1nduzKS/630-configuration-validation-for-broker-daemon)

File: package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "ci-test": "npm run format && npm test",
     "start-sparkswapd": "pm2-docker pm2.json --watch",
     "broker-proto": "pbjs -t proto3 broker.proto -p ./broker-daemon/proto -o /dev/null && cp ./broker-daemon/proto/broker.proto ./broker-cli/proto/broker.proto",
+    "validate-config": "standard ~/.sparkswap.js",
```
Comment (treygriffith) [inline package.json]: i don't think this actually validates the config...

File: broker-cli/broker-daemon-client/config.js [hunk 2]
```diff
@@ -15,6 +22,13 @@ try {
    */
   var USER_CONFIG = require(path.resolve(os.homedir(), '.sparkswap.js'))
 } catch (e) {
+  // We will only trigger an error if the `~/.sparkswap.js` configuration file
+  // exists on the user's machine. Otherwise we just silently set to a default
+  // configuration.
+  if (e.message && !e.message.includes(MISSING_FILE_MESSAGE)) {
```
Comment (treygriffith) [inline broker-cli/broker-daemon-client/config.js]: let's put this into a function so we can test it

File: package.json (hunk 1 above)
Comment (dannypaz) [reply package.json]: @treygriffith what is your hesitation? The screenshot in the description is an example of it working against a bad file

Comment (dannypaz) [reply package.json]: @treygriffith I think I understand what you mean. This is one type of validation, however ive added a card that can help us actually write a 'validator' for this (if rpcAddress exists, if whatever keys exist)

File: broker-cli/broker-daemon-client/config.js (hunk 2 above)
Comment (dannypaz) [reply broker-cli/broker-daemon-client/config.js]: @treygriffith added tests

File: broker-cli/broker-daemon-client/config.js [hunk 3]
```diff
@@ -1,25 +1,44 @@
...
+ *
+ * We attempt to load the file and set default configuration values. We will also
+ * warn the user if we have failed to parse their user config file.
+ *
+ * @return {Object} config
  */
-module.exports = Object.assign({}, DEFAULT_CONFIG, USER_CONFIG)
+function loadConfig () {
+  try {
+    var config = require(path.resolve(os.homedir(), '.sparkswap.js'))
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-cli/broker-daemon-client/config.js]: @treygriffith I think we talked about it, but I really like having the comments inside of the file. We would not be able to do that with JSON

File: broker-cli/broker-daemon-client/config.js [hunk 4]
```diff
@@ -1,25 +1,44 @@
...
+  } catch (e) {
+    // We will only trigger an error if the `~/.sparkswap.js` configuration file
+    // exists on the user's machine. Otherwise we just silently set to a default
```
Comment (martineehrlich) [inline broker-cli/broker-daemon-client/config.js]: We will warn the user only if the `~/.sparkswap.js` configuration file exists but could not be read. We will set the config to the default regardless of the error.

**Pull request merged**